docs: flesh out README.md (#59)
canonical/jubilant: README.md

# Jubilant, the joyful library for integration-testing Juju charms

![CI Status](https://github.com/canonical/jubilant/actions/workflows/ci.yaml/badge.svg)

> 

Reviewer: It's not important for this PR, but I am curious about your thoughts on badges -- I don't think we've ever talked about those. It feels like badges were super hot for a while (leading to situations like the ops-scenario README until very recently, or even more extreme cases), and are less so now.

These do get used by Jon's releasegen tool (although only the "operator" team repositories show up there, not the "charm tech" team ones), although I feel like the tool probably could get the data without needing the badge, particularly with some consistent naming.

I'm not really convinced of the value of badges, particularly test/build workflow ones. I'm not opposed to them, either (when used judiciously), just not sure what they are bringing. I'm curious about the motivation here.

Author: Yeah, this is a good call. I just copied this from Ops without much thought. But I agree with you -- I'm not sure what value it provides, and GitHub already shows a little green tick on the latest commit. Removed.

```
$ pip install jubilant
# or
$ uv add jubilant
```

To use Jubilant in Python code, instantiate a `Juju` instance and call methods like `deploy` and `wait`:

```python
import jubilant

juju = jubilant.Juju()
juju.deploy('snappass-test')
juju.wait(jubilant.all_active)
```

Below is an example integration test [from Jubilant itself](https://github.com/canonical/jubilant/blob/main/tests/integration/test_basic.py). This test uses the [`juju` pytest fixture](https://github.com/canonical/jubilant/blob/main/tests/integration/conftest.py), which adds a model during setup, runs your test with a `Juju` instance pointing at that model, and destroys the model during teardown:

```python
def test_deploy(juju: jubilant.Juju):
    juju.deploy('snappass-test')             # Deploy the charm
    status = juju.wait(jubilant.all_active)  # Wait till the app and unit are 'active'

    # Hit the Snappass HTTP endpoint to ensure it's up and running.
    address = status.apps['snappass-test'].units['snappass-test/0'].address
    response = requests.get(f'http://{address}:5000/', timeout=10)
    response.raise_for_status()
    assert 'snappass' in response.text.lower()
```

> You don't have to use [Pytest](https://docs.pytest.org/en/stable/) with Jubilant, but it's what we usually recommend. Pytest's fixtures and its simple `assert`-based approach make writing tests easy.

Reviewer: I feel like you could drop the "usually" here. It's probably more accurate, but I think for the README it invites unnecessary thought about what the unusual cases might be.

Author: Good call, done.

## Design goals

We designed Jubilant so it would:

- Match the Juju CLI. Method, parameter, and response field names match the Juju CLI, with minor exceptions (such as "application" being shortened to "app").
- Have a simple API. Higher-level operations will be in helpers and fixtures, not the main `Juju` class (the only exception being `Juju.wait`).

> - Not use `async`. This was a "feature" of python-libjuju that adds complexity and isn't needed for integration tests.

Reviewer: Is it worth mentioning here that most Juju CLI commands return quickly and complete asynchronously in the background anyway? I feel that's the main reason why using Python async was the wrong choice - for example if you asynchronously call `deploy` when that completes the deploy isn't complete, just whatever initial validation and scheduling is done; you need to do some sort of waiting for the deploy to really complete.

Author: Good call, done.

- Support Juju 3 and 4. The Juju team is guaranteeing CLI arguments and `--format=json` responses won't change between Juju 3.x and 4.x. When Juju 5.x arrives and changes the CLI, we'll keep the Jubilant API simple and 1:1 with the 5.x CLI, but will consider adding a `jubilant.compat` layer to avoid tests have to manually work around differences between 4.x and 5.x.

## Contributing and developing

Anyone can contribute to Jubilant. It's best to start by [opening an issue](https://github.com/canonical/jubilant/issues) with a clear description of the problem or feature request, but you can also [open a pull request](https://github.com/canonical/jubilant/pulls) directly.

> Jubilant uses [`uv`](https://docs.astral.sh/uv/) to manage Python dependencies and tools, so you'll need to [install uv](https://docs.astral.sh/uv/#installation) to work on the library. You'll also need `make` to run local development tasks (but you probably have `make` installed already).

Reviewer: You're almost certainly going to want to have Juju, I think. Do you feel that the link at the top to `juju.is` is sufficient for that? It does have a prominent link to the tutorial. I wonder if a link to [setting up a dev environment](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-your-deployment/manage-your-deployment-environment/#manage-your-deployment-environment) is worthwhile here as a helper? Or even something along the lines of "you can get a simple Juju environment using the 'charm-dev' Multipass blueprint: run `multipass launch ... charm-dev`".

Or is this all pointless because why are you looking at the Jubilant README if you don't already have Juju?

Author: True. I think this does somewhat go without saying. However, you also need it to *use* Jubilant, so I've added a paragraph with a link to the "install juju" page further up. They can follow further docs there if they really are new to Juju (which is highly unlikely if they want to use Jubilant).

After that, clone the Jubilant codebase and type `make all` to run various checks and the unit tests:

```
$ git clone https://github.com/canonical/jubilant
Cloning into 'jubilant'...
...
$ cd jubilant
$ make all
...
========== 107 passed in 0.26s ==========
```

To contribute a code change, write your fix or feature, add tests, document it, and run `make all` before you push and create a PR. Once you create a PR, GitHub will also run the integration tests, which takes several minutes.

## Doing a release

To create a new release of Jubilant:

- Update the `__version__` field in [`jubilant/__init__.py`](https://github.com/canonical/jubilant/blob/main/jubilant/__init__.py) to the new version you want to release.
- Push up a PR with this change and get it reviewed and merged.

> - Create a [new release](https://github.com/canonical/jubilant/releases/new) on GitHub with good release notes. The tag should start with a `v`, like `v1.2.3`. Once you've created the release, the [`publish.yaml` workflow](https://github.com/canonical/jubilant/blob/main/.github/workflows/publish.yaml) will automatically publish it to PyPI.

Reviewer: Do you want to explicitly say that it's semver here? More for people consuming than people publishing.

Author: Good point. However, I'm going to add this to the little "Releases" section on the docs landing page (PR coming soon).

It was written to supersede the use of [pytest-operator](https://github.com/charmed-kubernetes/pytest-operator) and [python-libjuju](https://github.com/juju/python-libjuju/) for charm integration tests. Python-libjuju in particular has a complex and confusing API, and its use of `async` is unnecessary for testing.

> Jubilant is currently in pre-release or "beta" phase ([PyPI releases](https://pypi.org/project/jubilant/#history)). Our intention is to release a 1.0.0 final version in May 2025.

Author: Done

> [**Read the full documentation.**](https://canonical-jubilant.readthedocs-hosted.com/)

Reviewer 2: Very minor suggestion:

```suggestion
[**Read the full documentation**](https://canonical-jubilant.readthedocs-hosted.com/)
```

Author: Done

## Using Jubilant

Jubilant is published to PyPI, so you can install and use it with your favourite Python package manager:

```
$ pip install jubilant
# or
$ uv add jubilant
```

Because Jubilant calls the Juju CLI, you'll also need to [install Juju](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-juju/#install-juju).

> To use Jubilant in Python code, instantiate a `Juju` instance and call methods like `deploy` and `wait`:

Reviewer 2: Matter of taste, but I feel this would be fine (no need to re-explain the code):

```suggestion
To use Jubilant in Python code:
```

Author: Good call, done

```python
import jubilant

juju = jubilant.Juju()
juju.deploy('snappass-test')
juju.wait(jubilant.all_active)
```

Below is an example integration test [from Jubilant itself](https://github.com/canonical/jubilant/blob/main/tests/integration/test_basic.py). This test uses the [`juju` pytest fixture](https://github.com/canonical/jubilant/blob/main/tests/integration/conftest.py), which adds a model during setup, runs your test with a `Juju` instance pointing at that model, and destroys the model during teardown:

```python
def test_deploy(juju: jubilant.Juju):
    juju.deploy('snappass-test')             # Deploy the charm
    status = juju.wait(jubilant.all_active)  # Wait till the app and unit are 'active'

    # Hit the Snappass HTTP endpoint to ensure it's up and running.
    address = status.apps['snappass-test'].units['snappass-test/0'].address
    response = requests.get(f'http://{address}:5000/', timeout=10)
    response.raise_for_status()
    assert 'snappass' in response.text.lower()
```

> You don't have to use [Pytest](https://docs.pytest.org/en/stable/) with Jubilant, but it's what we recommend. Pytest's fixtures and its simple `assert`-based approach make writing tests easy.

Reviewer 2: We mention fixtures, but I don't see fixtures used in https://github.com/canonical/jubilant/blob/main/tests/integration/test_basic.py. And although we promise that it's easy to use fixtures in the context of Jubilant, I think an example would be really helpful.

How about something like this, with the final sentence added later perhaps:

```suggestion
You don't have to use [Pytest](https://docs.pytest.org/en/stable/) with Jubilant, but it's what we recommend. Pytest's `assert`-based approach is a straightforward way to write tests, and its fixtures are helpful for structuring more complex tests. For an example of how to use Pytest fixtures with Jubilant, see LINK.
```

> You don't have to use [Pytest](https://docs.pytest.org/en/stable/) with Jubilant, but it's what we recommend. Pytest's fixtures and its simple `assert`-based approach make writing tests easy.

Author: Yeah, I think you're right it was a bit confusing. However, I realised it wasn't actually helpful to mention Jubilant's integration tests -- this happens to come from one of the Jubilant integration tests, but that's unimportant, and adds to the confusion. I've reworked this section a bit in a way that hopefully clarifies.

## Design goals

We designed Jubilant so it would:

- Match the Juju CLI. Method, parameter, and response field names match the Juju CLI, with minor exceptions (such as "application" being shortened to "app").
- Have a simple API. Higher-level operations will be in helpers and fixtures, not the main `Juju` class (the only exception being `Juju.wait`).
- Not use `async`. This was a "feature" of python-libjuju that adds complexity and isn't needed for integration tests. In addition, most Juju CLI commands return quickly and complete asynchronously in the background.

> - Support Juju 3 and 4. The Juju team is guaranteeing CLI arguments and `--format=json` responses won't change between Juju 3.x and 4.x. When Juju 5.x arrives and changes the CLI, we'll keep the Jubilant API simple and 1:1 with the 5.x CLI, but will consider adding a `jubilant.compat` layer to avoid tests have to manually work around differences between 4.x and 5.x.

Reviewer 2: I think it's fine to be less specific about future details. (Also a couple of readability nits.) How about:

```suggestion
- Support Juju 3 and 4. The Juju team is guaranteeing CLI arguments and `--format=json` responses won't change between Juju 3.x and 4.x. When Juju 5.x arrives and changes the CLI, we'll keep the Jubilant API simple and match the 5.x CLI. However, we will consider adding a compatibility layer to avoid tests having to manually handle differences between 4.x and 5.x.
```

Author: Done.

## Contributing and developing

Anyone can contribute to Jubilant. It's best to start by [opening an issue](https://github.com/canonical/jubilant/issues) with a clear description of the problem or feature request, but you can also [open a pull request](https://github.com/canonical/jubilant/pulls) directly.

Jubilant uses [`uv`](https://docs.astral.sh/uv/) to manage Python dependencies and tools, so you'll need to [install uv](https://docs.astral.sh/uv/#installation) to work on the library. You'll also need `make` to run local development tasks (but you probably have `make` installed already).

> After that, clone the Jubilant codebase and type `make all` to run various checks and the unit tests:

Reviewer 2: Very minor nit:

```suggestion
After that, clone the Jubilant codebase and use `make all` to run various checks and the unit tests:
```

Author: Done

```
$ git clone https://github.com/canonical/jubilant
Cloning into 'jubilant'...
...
$ cd jubilant
$ make all
...
========== 107 passed in 0.26s ==========
```

> To contribute a code change, write your fix or feature, add tests, document it, and run `make all` before you push and create a PR. Once you create a PR, GitHub will also run the integration tests, which takes several minutes.

Reviewer 2: Combine tests and docs for easier reading?

```suggestion
To contribute a code change, write your fix or feature, add tests and docs, then run `make all` before you push and create a PR. Once you create a PR, GitHub will also run the integration tests, which takes several minutes.
```

Author: Done

## Doing a release

To create a new release of Jubilant:

> - Update the `__version__` field in [`jubilant/__init__.py`](https://github.com/canonical/jubilant/blob/main/jubilant/__init__.py) to the new version you want to release.

Reviewer 2: Minor, but these should be numbered steps

Author: Done
